sdks/python/apache_beam/testing/load_tests/load_test_metrics_utils.py

```diff
@@ -81,120 +83,158 @@ def get_element_by_schema(schema_name, insert_list):
-    if len(outputs) > 0:
-      for output in outputs:
-        errors = output['errors']
-        for err in errors:
-          logging.error(err['message'])
-          raise ValueError(
-              'Unable save rows in BigQuery: {}'.format(err['message']))
+class MetricsReader(object):
+  publishers = []
 
-  def _get_dataset(self, dataset_name):
-    bq_dataset_ref = self._bq_client.dataset(dataset_name)
-    try:
-      bq_dataset = self._bq_client.get_dataset(bq_dataset_ref)
-    except NotFound:
-      raise ValueError(
-          'Dataset {} does not exist in your project. '
-          'You have to create table first.'
-          .format(dataset_name))
-    return bq_dataset
+  def __init__(self, project_name=None, bq_table=None, bq_dataset=None):
+    self.publishers.append(ConsoleMetricsPublisher())
+    check = project_name and bq_table and bq_dataset \
+            is not None
+    if check:
+      bq_publisher = BigQueryMetricsPublisher(
+          project_name, bq_table, bq_dataset)
+      self.publishers.append(bq_publisher)
 
-  def _get_or_create_table(self, bq_schemas, dataset):
-    if self._namespace == '':
-      raise ValueError('Namespace cannot be empty.')
+  def publish_metrics(self, result):
+    metrics = result.metrics().query()
+    insert_dicts = self._prepare_all_metrics(metrics)
+    for publisher in self.publishers:
+      publisher.publish(insert_dicts)
 
-    dataset = self._get_dataset(dataset)
-    table_ref = dataset.table(self._namespace)
+  def _prepare_all_metrics(self, metrics):
+    submit_timestamp = time.time()
+    metric_id = uuid.uuid4().hex
 
-    try:
-      self._bq_table = self._bq_client.get_table(table_ref)
-    except NotFound:
-      table = bigquery.Table(table_ref, schema=bq_schemas)
-      self._bq_table = self._bq_client.create_table(table)
+    insert_rows = []
 
-  def _prepare_schema(self):
-    return [SchemaField(**row) for row in SCHEMA]
+    for counter in metrics['counters']:
+      counter_dict = CounterMetric(counter, submit_timestamp, metric_id)\
+        .as_dict()
+      insert_rows.append(counter_dict)
 
-  def _get_schema_names(self):
-    return [schema['name'] for schema in SCHEMA]
+    dists = metrics['distributions']
+    if len(dists) > 0:
+      runtime = RuntimeMetric(dists, submit_timestamp, metric_id)\
+        .as_dict()
+      insert_rows.append(runtime)
 
+    return insert_rows
 
-class MetricsMonitor(object):
-  def __init__(self, project_name, table, dataset):
-    if project_name is not None:
-      self.bq = BigQueryClient(project_name, table, dataset)
 
-  def send_metrics(self, result):
-    metrics = result.metrics().query()
+class Metric(object):
+  value = None
+  label = None
 
-    insert_dicts = self._prepare_all_metrics(metrics)
+  def __init__(self, submit_timestamp, metric_id):
+    self.submit_timestamp = submit_timestamp
+    self.metric_id = metric_id
 
-    self.bq.match_and_save(insert_dicts)
+  def as_dict(self):
```

Comment:
I decided to create a method that will return object represented as a dict but I also thought about overloading `__iter__` method, so then the created object can be mapped like this: `dict(Metric(...))`. @DariuszAniszewski what is your opinion? Which would be better?

Reply:
If I understand correctly, instance of `Metric` represents a single metric reading. Thus `Metric` isn't really a iterable and `as_dict` makes more sense to me.

Reply:
Great, thanks for your opinion!